race-buffer/src/buffer.rs

```diff
@@ -0,0 +1,635 @@
+//! The writer of the RaceBuffer, which contains the actual buffer. Entries can be synchronously written and read/iterated
+//! from this struct.
+use crate::{get_seqn_index, num_missed, Entry, SeqNum, WholeEntry};
+//use core::iter::Iterator;
+use core::cmp::max;
+use core::fmt;
+use core::mem::size_of;
+use core::mem::MaybeUninit;
+use core::sync::atomic::fence;
+use core::sync::atomic::Ordering;
+
+/// Minimum allowed capacity of backing storage
+pub const MIN_STORAGE_CAP: usize = 4;
+
+/// Error for given storage size that is too small
+pub struct SizeError;
+
+impl fmt::Debug for SizeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Storage capacity must be at least {}", MIN_STORAGE_CAP,)
+    }
+}
+
+/// Round given length down to a power of 2
+#[inline]
+fn round_to_power_2(length: usize) -> usize {
```

Comment:
This function panics when given a value of 0. Probably ought to fix that, and add tests.

Reply:
I added a note and debug assertion. This function is tested in `buffer::tests::test_init_sizes`